src/compiler/checker.ts

```diff
@@ -20343,15 +20349,22 @@ namespace ts {
                 if (type.flags & TypeFlags.Any && literal.text === "function") {
                     return type;
                 }
+                if (assumeTrue && type.flags & TypeFlags.Unknown && literal.text === "object") {
+                    // The pattern x && typeof x === 'object', where x is of type unknown, narrows x to type object.

```

Comment:
Would be nice to note here that you don't need to check both directions because the `typeof x === "object" && x` is narrowed as two separate steps. I don't think that would be obvious to a casual reader.